### PR TITLE
feat(Auth): Add unit tests for TOTP

### PR DESCRIPTION
**Pull request by harsh62**
### TOTP PR Number # 7

Depends on: 
* #3040 
* #3041 
* #3042
* #3043 
* #3044 
* #3045 

Changes
* Adds all the unit tests

_NOTE: The unit and integration tests will fail until all changes make it to the feature branch_
(The PR is directed towards a feature branch)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by ruisebas — APPROVED** [review]

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/MFA/FetchMFAPreferenceTaskTests.swift [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+
+    /// Test a successful fetchMFAPreference call
+    ///
+    /// - Given: an auth plugin with mocked service. Mocked service calls should mock a successul response
+    /// - When:
+    ///    - I invoke fetchMFAPreference
+    /// - Then:
+    ///    - I should get a successful result
+    ///
+    func testSuccessfulPreferenceFetchWithTOTPPreferred() async {
```
Comment (royjit) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/MFA/FetchMFAPreferenceTaskTests.swift:27]: Should we add a test where service returns a new MFA type or invalid strings?

**Review by royjit — APPROVED** [review]

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/MFA/FetchMFAPreferenceTaskTests.swift [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+@testable import AWSCognitoAuthPlugin
+import AWSCognitoIdentityProvider
+
+// swiftlint:disable type_body_length
+// swiftlint:disable file_length
+class FetchMFAPreferenceTaskTests: BasePluginTest {
+
+    /// Test a successful fetchMFAPreference call
+    ///
+    /// - Given: an auth plugin with mocked service. Mocked service calls should mock a successul response
```
Comment (thisisabhash) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/MFA/FetchMFAPreferenceTaskTests.swift:21]: nit typo: successful
similarly in tests below

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/MFA/VerifyTOTPSetupTaskTests.swift [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+            let pluginOptions = VerifyTOTPSetupOptions(friendlyDeviceName: "device")
+            try await plugin.verifyTOTPSetup(
+                code: "123456", options: .init(pluginOptions: pluginOptions))
+        } catch {
+            XCTFail("Received failure with error \(error)")
+        }
+    }
+
+
+//case invalidUserPoolConfigurationException(InvalidUserPoolConfigurationException)
```
Comment (thisisabhash) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/MFA/VerifyTOTPSetupTaskTests.swift:46]: nit: remove comment

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithSetUpMFATaskTests.swift [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+                        secretCode: "sharedSecret",
+                        session: "session",
+                        username: "username")),
+                    .testData)),
+            AuthorizationState.sessionEstablished(.testData))
+    }
+
+    /// Test a successful confirmSignIn call with .done as next step
+    ///
+    /// - Given: an auth plugin with mocked service. Mocked service calls should mock a successul response
```
Comment (thisisabhash) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithSetUpMFATaskTests.swift:33]: nit typo: successful
similarly in tests below

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithSetUpMFATaskTests.swift [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+            XCTFail("Should not succeed")
+        } catch {
+            guard case AuthError.validation = error else {
+                XCTFail("Should produce validation error instead of \(error)")
+                return
+            }
+        }
+    }
+
+    /// Test a confirmSignIn call with an empty confirmation code followed by a second valaid confirmSignIn call
```
Comment (thisisabhash) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithSetUpMFATaskTests.swift:97]: nit typo: valid

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithSetUpMFATaskTests.swift [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+                return
+            }
+            guard case .codeMismatch = (underlyingError as? AWSCognitoAuthError) else {
+                XCTFail("Underlying error should be codeMismatch \(error)")
+                return
+            }
+        }
+    }
+
+    func testConfirmSignInRetryWithCodeMismatchException() async {
```
Comment (thisisabhash) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithSetUpMFATaskTests.swift:175]: nit : given-when-then doc

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithMFASelectionTaskTests.swift [hunk 7]
```diff
@@ -0,0 +1,836 @@
...
+                        .apiBased(.userSRP)
+                    ),
+                    .selectMFAType,
+                    .apiBased(.userSRP))),
+            AuthorizationState.sessionEstablished(.testData))
+    }
+
+    /// Test a successful confirmSignIn call with .done as next step
+    ///
+    /// - Given: an auth plugin with mocked service. Mocked service calls should mock a successul response
```
Comment (thisisabhash) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithMFASelectionTaskTests.swift:34]: nit: typo successful
similar across all files

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithMFASelectionTaskTests.swift [hunk 8]
```diff
@@ -0,0 +1,836 @@
...
+                return
+            }
+            guard case .codeMismatch = (underlyingError as? AWSCognitoAuthError) else {
+                XCTFail("Underlying error should be codeMismatch \(error)")
+                return
+            }
+        }
+    }
+
+    func testConfirmSignInRetryWithCodeMismatchException() async {
```
Comment (thisisabhash) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithMFASelectionTaskTests.swift:253]: nit: given-when-then doc

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithMFASelectionTaskTests.swift [hunk 9]
```diff
@@ -0,0 +1,836 @@
...
+    }
+
+    /// Test a confirmSignIn call with InvalidSmsRoleAccessPolicy response from service
+    ///
+    /// - Given: an auth plugin with mocked service. Mocked service should mock a
+    ///   InvalidSmsRoleAccessPolicyException response
+    /// - When:
+    ///    - I invoke confirmSignIn with a valid confirmation code
+    /// - Then:
+    ///    - I should get a --
```
Comment (thisisabhash) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithMFASelectionTaskTests.swift:450]: nit: maybe there's something missing here about the expected result

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithMFASelectionTaskTests.swift [hunk 10]
```diff
@@ -0,0 +1,836 @@
...
+    }
+
+    /// Test a confirmSignIn call with InvalidSmsRoleTrustRelationship response from service
+    ///
+    /// - Given: Given an auth plugin with mocked service. Mocked service should mock a
+    ///   CodeDeliveryFailureException response
+    /// - When:
+    ///    - I invoke confirmSignIn with a valid confirmation code
+    /// - Then:
+    ///    - I should get a --
```
Comment (thisisabhash) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInWithMFASelectionTaskTests.swift:481]: similar comment as above

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInTOTPTaskTests.swift [hunk 11]
```diff
@@ -0,0 +1,806 @@
...
+    }
+
+    /// Test a confirmSignIn call with InvalidSmsRoleTrustRelationship response from service
+    ///
+    /// - Given: Given an auth plugin with mocked service. Mocked service should mock a
+    ///   CodeDeliveryFailureException response
+    /// - When:
+    ///    - I invoke confirmSignIn with a valid confirmation code
+    /// - Then:
+    ///    - I should get a --
```
Comment (thisisabhash) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/SignIn/ConfirmSignInTOTPTaskTests.swift:451]: similar comment as in the other file

**Comment by harsh62** [comment]
@thisisabhash I will try to resolve these comments in a future PR.

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/MFA/FetchMFAPreferenceTaskTests.swift (hunk 1 above)
Comment (harsh62) [reply AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TaskTests/ClientBehaviorTests/MFA/FetchMFAPreferenceTaskTests.swift:27]: Sure.. Will add it in a follow up PR.

**Pull request merged**